### ISSUE TITLE
Investigate why we get (maybe) good performance on encoding task with a randomly initialized model

### ISSUE THREAD

**Issue by zparis99** [leoniekerken/ECoG-foundation-model#27]
When testing the new colab with a barely trained model I was getting pearson correlations > 0.98 and a low mspe ~0.11. Why is that?

I would've expected our neural embeddings to essentially be noise and have the following guesses as to what's going on:

1. There's a bug in the encoding code which made it look better than it was?
2. I'm using the metrics for encoding incorrectly (there's another issue to investigate that more). What direction along embeddings should correlations be calculated? MSPE fitting?
3. The numbers I was seeing weren't actually good performance.
4. I got really really really lucky.
5. Something else?

**Comment by zparis99** [comment]
Looking more into this.

To document this more using a randomly initialized model leads to the following performance: [mean correlation: 0.993, min correlation: 0.916, max correlation: 0.999]. This is surprising because I would expect an untrained model to output random embeddings which would contain little information for the model to learn.

The histogram of correlations looks like this so clearly all at the high end:
<img width="603" alt="image" src="https://github.com/user-attachments/assets/2be8f1ef-9892-491a-9420-f41ee20133ef">

We can also see in individual cases how correlated the embeddings are. This is one example embedding:

<img width="639" alt="image" src="https://github.com/user-attachments/assets/2ec668a3-9378-4b84-936f-3be044f42f59">

To make sure this isn't some issue with the correlation we can also check the more often used cosine similarity although as this paper mentions: https://aclanthology.org/N19-1100.pdf cosine similarity is just pearson correlation without subtracting the mean so we would expect the same result here. For this we get practically the same results: [mean: 0.993, min: 0.916, max: 0.999]

To make sure that the assumption about random embeddings leading to poor performance is true I tried first to see what kind of performance we get with fully randomized "neural" embeddings. By uniformly sampling between [0, 1] we get the following results: [mean: 0.001, min:-0.38, max: 0.39] which is more in-line with our expectations.

**Comment by zparis99** [comment]
To check if this had something to do with the distribution of the output embeddings I ran a Shapiro-Wilk test on all of the neural embeddings and found that the p-values were all significantly too high to deny the hypothesis that the data is normally distributed suggesting that it is normally distributed. To test if this has an impact I ran similar analysis to above with normally distributed randomly sampled data and didn't find any clear correlations: [mean: 0.001, max: 0.409, min: -0.405] so no dice there.

**Comment by zparis99** [comment]
To confirm that there is something meaningful in the ordering of the dimensions of the embeddings themselves I tried shuffling the neural embeddings along their embedding axis. This leads to: mean: -0.002, min: -0.410, max: 0.376 so I think that it's safe to say that there is something going on with these embeddings. I'm going to search and see if there are any bugs in the dataloader that are making this happen. I'm very skeptical that the weights in the model are at all meaningful so it seems more likely to me that there is something which makes the encoding task easy.

**Comment by zparis99** [comment]
So I think I've maybe figured it out. This was pretty fun to find. Currently the dimensionality of the embeddings is set to 80. If I run PCA over the output neural embeddings of the random model 85% of the variance is explained by 3 dimensions. This suggests that the model is just outputting a 3d-ish manifold with some added noise which has been rotated into 80 dimensions. What I think is happening then is that the encoding model is just learning to output random points in this 3d manifold and the comparatively higher dimensionality of these embeddings means that that is a good enough guess to get a high correlation.

To test this I did exactly this. I drew points from a 3d normal distribution (normal because this is what the histogram of the PCA'd neural embeddings looked like: 
<img width="583" alt="image" src="https://github.com/user-attachments/assets/dd648b10-1d1e-4205-8a1a-208acf36a15b">
).

I then rotated all of these points into an 80-dimensionaal space using a random rotation matrix and added some random noise to until the variance explained by 3-dimensions in PCA was equivalent to that of the real data. I then ran regression and got the following: mean: -0.004, min: -0.946, max: 0.954, min abs value: 0.000, abs mean: 0.428. While this is not equivalent to the original case we can see that the encoding model is able to on-average find a higher correlation than the random data even if it is not quite as high as the original case. I'll look more into trying to confirm this tomorrow, but I believe this is likely what's happening here.

**Comment by zparis99** [comment]
So this is actually a very simple problem, but will be helpful I think for when we try to determine the quality of our embeddings. 

I did some more testing where I generate random uniformly distributed points and then rotate them into 80 dimensions with some random rotation matrix. I then add random noise AND add the mean of each dimension in the neural embedding data so it is no longer 0-centered. This then leads to very similar performance to the actual model output data. mean: 0.994, min: 0.898, max: 1.000, min abs value: 0.898, abs mean: 0.994.

The addition of the mean then is what led to the encoding task consistently performing so well. If we look at the mean vector of the neural embeddings we see that it has a norm of 7.5 which is greater than the variance of the embeddings themselves. The result of this is easy to see in a reduced dimension view of the embeddings:

<img width="554" alt="image" src="https://github.com/user-attachments/assets/d8351965-f723-4c8d-998f-500270325b9e">

Basically the model is outputting a normal distribution centered around some non-zero point, so the encoding model learns to predict that distribution. Because the points are far enough away from the origin it trivially achieves a high encoding performance as measured through cosine similarity, but because pearson correlation is supposed to subtract the mean I was surprised to see this. This revealed a bug in the code because it was subtracting the means from the wrong dimension, ugh.

**Comment by zparis99** [comment]
On the bright side this does suggest some other things we should make sure to keep track of when doing the encoding performance checks:

1. Make sure that we get low explained variance using PCA of the embeddings over low-dimensions. If we do then we should check that the model is not outputting trivially simple embeddings.
2. Make sure the mean of the pairwise cosine similarity between all of the embeddings is not too high. This suggests that all of the points are output on the same line relative to the origin.
3. Include mean squared prediction error in the analysis as well. It's not always super meaningful in high dimensions but could help catch corner cases like this in our pearson measure.

**Comment by zparis99** [comment]
As per meeting the other solution to this as far as testing our model is that we won't getting similarly good performance with a decoding task. Done implementing that and will test soon.
